src/mewbot/io/desktop_notification.py

```diff
@@ -0,0 +1,242 @@
+    _platform_str: str
+    _enabled: bool
+
+    def __init__(self) -> None:
+
+        self._logger = logging.getLogger(__name__ + "DesktopNotificationOutputEngine")
+        self._logger.info("DesktopNotificationOutputEngine starting")
+
+        # Note-to-self: Warn if we are in a non-tested state
+        # Note-to-self: Need to test to see if we are headless and fallback if we are
+
+        self._platform_str = sys.platform
+        self._logger.info(f"We are detected as running on {self._platform_str}")
+
+        # Could use "in" or string normalisation - but I want it to fail noisily and log the error
+        # if I have made any incorrect assumptions.
+        # So, hopefully, the user will tell us and I can check to see if anything creative
+        # is required.
+        # silent failure of desktop notifications is not desired. Noisy failure is.
+
+        self._enabled = False
+        if self._platform_str == "win32" or self._platform_str == "win64":
+            self._detected_os = "windows"
+            self._do_windows_setup()
+
+        elif self._platform_str == "linux":
+            if self._platform_str == "linux2":
+                self._logger.warning(
+                    f"You seem to be running python below 3.3, "
+                    f"or python behavior has changed - sys.platform = {self._platform_str}"
+                )
+
+            self._detected_os = "linux"
+            setattr(self, "_notify", self._linux_notify_send_method)
+
+        elif self._platform_str == "darwin":
+            self._detected_os = "macos"
+            self._enabled = False
+
+        elif "freebsd" in self._platform_str or "dragonfly" in self._platform_str:
+            self._logger.warning(
+                f"untested configuration - {self._platform_str} - attempting freebsd like behavior"
+            )
+            self._detected_os = "freebsd"
+            self._enabled = False
+
+        elif "haiku" in self._platform_str:
+            self._logger.warning(
+                f"untested configuration - {self._platform_str} - attempting haiku like behavior"
+            )
+            self._detected_os = "haiku"
+            self._enabled = False
+
+        else:
+            self._logger.warning(
+                f"Unsupported configuration {self._platform_str} - cannot enable"
+            )
+            self._enabled = False
+
+    def notify(self, title: str, text: str) -> bool:
+        """
+        Preform the actual notification task using the internal setup.
+        Means we can avoid live patching externally accessible functions.
+        """
+        if not self._enabled:
+            return False
+        return self._notify(title, text)
+
+    def _notify(self, title: str, text: str) -> bool:
+        self._logger.warning(
+            f"{self._notify} should never be called directly - "
+            f"should have been live replaced with the true method - f{title} - f{text}"
+        )
+        return False
```

Comment:
An alternative to this would be a dict of [detected OS; function to call] which might be easier to follow:
```suggestion
    def _notify(self, title: str, text: str) -> bool:
        # Probably don't do this inline...
        caller = {"windows": self._windows_toast_method, "linux": self._linux_notify_send_method}.get(self._detected_os)
        
        if not caller:
          self._logger.warning("No notification method supported for %s", self._detected_os)
          return False
          
        return caller(title, text)
```

Reply:
Does the setattr method appreciably reduce runtime?

Reply:
`setattr` will be faster than what I've outlined here (as there won't essentially be a proxy to another method) and is almost certainly the fastest method for live matching something in standard python. On the flip side, if there's enough notifications that an extra layer of stack indirection makes a difference to real world performance, there's going to be so many notifications

Reply:
It is easier to read - correcting.